### PR TITLE
Add location from user and default group from user and first group fromuser for rule asset

### PR DESCRIPTION
**Pull request by ales-infotel**
This is a new feature for assets rules. It adds  new criterias User location and user in a group. And a action to copy location from user and a action to copy group default from user.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by ales-infotel** [comment]
I don't understand why my PR generates an error that is not related to the files I modified and the other files come from the master branch.
see test failure :
> There is 1 failure:
=> tests\units\Glpi\Csv\LogCsvExport::testGetFileName():
In file /var/glpi/tests/units/Glpi/Csv/LogCsvExport.php on line 0, string() failed for data set [0] of data provider csvTestProvider: strings are not equal
-Expected
+Actual
@@ -1 +1 @@
-string(29) "testexporttocsv-520210211.csv"
+string(32) "testexporttocsv-5_2021_02_11.csv"

**Comment by cedric-anne** [comment]
> I don't understand why my PR generates an error that is not related to the files I modified and the other files come from the master branch.
> see test failure :
> 
> > There is 1 failure:
> > => tests\units\Glpi\Csv\LogCsvExport::testGetFileName():
> > In file /var/glpi/tests/units/Glpi/Csv/LogCsvExport.php on line 0, string() failed for data set [0] of data provider csvTestProvider: strings are not equal
> > -Expected
> > +Actual
> > @@ -1 +1 @@
> > -string(29) "testexporttocsv-520210211.csv"
> > +string(32) "testexporttocsv-5_2021_02_11.csv"

Hi,

This has been fixed on master. Please rebase on last commit to be sure that other tests are passing.

Regards

**Comment by ales-infotel** [comment]
ok thank, now the tests are good

**Review by cedric-anne — COMMENTED** [review]
It seems correct.

Please add some unit tests to validate this behaviour.

**Comment by trasher** [comment]
Looks good to me, waiting for tests to validate

**Comment by ales-infotel** [comment]
i have add somes tests

**Review by trasher — APPROVED** [review]

File: inc/commondbtm.class.php [hunk 1]
```diff
@@ -5413,6 +5413,19 @@ private function assetBusinessRules($condition) {
...
 
+         $user = new User();
+         if (isset($input["users_id"]) && $input["users_id"] != 0
+             && $user->getFromDB($input["users_id"])) {
+            $group_user  = new Group_User();
+            $groups_user = $group_user->find(['users_id' => $input["users_id"]]);
+            foreach ($groups_user as $group) {
+               $input['_groups_id_of_user'] = $group['groups_id'];
+               break;
+            }
```
Comment (cedric-anne) [inline inc/commondbtm.class.php:5424]:
```suggestion
            $input['_groups_id_of_user'] = [];
            foreach ($groups_user as $group) {
               $input['_groups_id_of_user'][] = $group['groups_id'];
            }
```

File: inc/ruleasset.class.php [hunk 2]
```diff
@@ -232,6 +246,27 @@ function executeActions($output, $params, array $input = []) {
...
+                  if (($action->fields['field'] == 'groups_id')
+                      && isset($input['_default_groups_id_of_user'])) {
+                     $output['groups_id'] = $input['_default_groups_id_of_user'];
+                  }
+                  break;
+
+               case 'firstgroupfromuser':
+                  if (($action->fields['field'] == 'groups_id')
+                      && isset($input['_groups_id_of_user'])) {
+                     $output['groups_id'] = $input['_groups_id_of_user'];
```
Comment (cedric-anne) [inline inc/ruleasset.class.php:266]:
```suggestion
                     $output['groups_id'] = reset($input['_groups_id_of_user']);
```

File: tests/units/RuleAsset.php [hunk 3]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+
+      // Test on creation
+      $computer_em = new \Computer();
+      $computer_id = $computer_em->add([
+                                          'name'        => 'test',
+                                          'entities_id' => 0,
+                                          'users_id'    => $users_id,
+                                       ]);
+
+      $computer_em->getFromDB($computer_id);
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:313]:
```suggestion
      $this->integer($computer_id)->isGreaterThan(0);
      $this->boolean($computer_em->getFromDB($computer_id))->isTrue();
```

File: tests/units/RuleAsset.php [hunk 4]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+
+      // Test on update
+      $computer_em = new \Computer();
+      $computer_id = $computer_em->add([
+                                          'name'        => 'test2',
+                                          'entities_id' => 0,
+                                       ]);
+      $this->integer($computer_id)->isGreaterThan(0);
+
+      $computer_em->getFromDB($computer_id);
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:324]:
```suggestion
      $this->boolean($computer_em->getFromDB($computer_id))->isTrue();
```

File: tests/units/RuleAsset.php [hunk 5]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+      $this->integer($computer_em->getField('locations_id'))->isNotEqualTo($locations_id);
+
+      $computer_em->update([
+                              'id'       => $computer_id,
+                              'users_id' => $users_id,
+                           ]);
+
+      $computer_em->getFromDB($computer_id);
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:332]:
```suggestion
      $update = $computer_em->update([
                              'id'       => $computer_id,
                              'users_id' => $users_id,
                           ]);
      $this->boolean($update)->isTrue();
      $this->boolean($computer_em->getFromDB($computer_id))->isTrue();
```

File: tests/units/RuleAsset.php [hunk 6]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+      $group    = new \Group();
+      $group_id = $group->add($group_input = [
+         "name"         => "group1",
+         "is_requester" => true
+      ]);
+      $this->checkInput($group, $group_id, $group_input);
+
+      //Load user tech
+      $user = new \User();
+      $user->getFromDB(getItemByTypeName('User', 'tech', true));
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:382]:
```suggestion
      $user = getItemByTypeName('User', 'tech');
```

File: tests/units/RuleAsset.php [hunk 7]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+
+      // Check ticket that trigger rule on creation
+      $computer     = new \Computer();
+      $computers_id = $computer->add($computer_input = [
+         'name'        => 'test',
+         'entities_id' => 0,
+         'users_id'    => $user->fields['id']
+      ]);
+
+      $computer->getFromDB($computers_id);
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:404]:
```suggestion
      $this->integer($computers_id)->isGreaterThan(0);
      $this->boolean($computer->getFromDB($computers_id))->isTrue();
```

File: tests/units/RuleAsset.php [hunk 8]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+      $group2    = new \Group();
+      $group_id2 = $group2->add($group_input2 = [
+         "name"         => "group2",
+         "is_requester" => true
+      ]);
+      $this->checkInput($group2, $group_id2, $group_input2);
+
+      //Load user tech
+      $user = new \User();
+      $user->getFromDB(getItemByTypeName('User', 'tech', true));
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:462]:
```suggestion
      $user = getItemByTypeName('User', 'tech');
```

File: tests/units/RuleAsset.php [hunk 9]
```diff
@@ -254,4 +254,237 @@ private function _createRuleLocation($condition) {
...
+
+      // Check ticket that trigger rule on creation
+      $computer     = new \Computer();
+      $computers_id = $computer->add($computer_input = [
+         'name'        => 'test',
+         'entities_id' => 0,
+         'users_id'    => $user->fields['id']
+      ]);
+
+      $computer->getFromDB($computers_id);
```
Comment (cedric-anne) [inline tests/units/RuleAsset.php:486]:
```suggestion
      $this->integer($computers_id)->isGreaterThan(0);
      $this->boolean($computer->getFromDB($computers_id))->isTrue();
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**